File: 卡码/代码随想录/Java/src/com/kurumi/codeThoughts/图论/dfs/LeetCode_0695.java
package com.kurumi.codeThoughts.图论.dfs;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 695. 岛屿的最大面积
 */
public class LeetCode_0695 {
    static class Solution {
        int[][] dir = new int[][] { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };
        int cnt = 0;

        public void dfs(int[][] grid, int x, int y) {
            grid[x][y] = 0;
            for (int i = 0; i < 4; i++) {
                int nextX = x + dir[i][0];
                int nextY = y + dir[i][1];
                if (nextX < 0 || nextX > grid.length - 1 || nextY < 0 || nextY > grid[0].length - 1)
                    continue;
                if (grid[nextX][nextY] == 1) {
                    cnt++;
                    dfs(grid, nextX, nextY);
                }
            }
        }

        public int maxAreaOfIsland(int[][] grid) {
            int n = grid.length;
            int m = grid[0].length;
            int res = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (grid[i][j] == 1) {
                        cnt = 1;
                        dfs(grid, i, j);
                        res = Math.max(res, cnt);
                    }
                }
            }
            return res;
        }
    }
}

File: 卡码/代码随想录/JS/图论/dfs/LeetCode_0695.js
//695. 岛屿的最大面积
const maxAreaOfIsland = function(grid) {
	const n=grid.length;
	const m=grid[0].length;
	let cnt=1,res=0;
	const dir=[[0,1],[1,0],[0,-1],[-1,0]];
	const dfs=(x,y)=>{
		grid[x][y]=0;
		for(let i=0;i<4;i++){
			const nextX=x+dir[i][0];
			const nextY=y+dir[i][1];
			if(nextX<0||nextX>n-1||nextY<0||nextY>m-1) continue;
			if(grid[nextX][nextY]==1){
				cnt++;
				dfs(nextX,nextY);
			}
		}
	}
	for(let i=0;i<n;i++){
		for(let j=0;j<m;j++){
			if(grid[i][j]==1){
				cnt=1;
				dfs(i,j);
				res=Math.max(res,cnt);
			}
		}
	}
	return res;
};
